File: packages/modules/narduk-logging/src/sanitize.ts
import type { JsonValue, LogError, LogRecord, SafeFields } from './types.js'

export const MAX_RECORD_BYTES = 16 * 1024
const MAX_STRING = 2048
const MAX_FIELDS = 50
const MAX_DEPTH = 6
const MAX_NODES = 500
const PRIVATE = Symbol('narduk.logging.private')
const encoder = new TextEncoder()
const REDACTED = '[REDACTED]'
const SENSITIVE = new Set([
  'password',
  'passwd',
  'secret',
  'token',
  'apikey',
  'authorization',
  'proxyauthorization',
  'cookie',
  'cookies',
  'setcookie',
  'session',
  'sessionid',
  'privatekey',
  'clientsecret',
  'body',
  'requestbody',
  'responsebody',
  'payload',
  'payment',
  'cardnumber',
  'cvv',
  'email',
  'phone',
  'address',
  'latitude',
  'longitude',
  'prompt',
  'completion',
])
/** Exact segments after camelCase / snake_case / kebab-case split. */
const SENSITIVE_SEGMENTS = new Set([
  'password',
  'passwd',
  'secret',
  'token',
  'apikey',
  'authorization',
  'cookie',
  'cookies',
  'setcookie',
  'session',
  'sessionid',
  'privatekey',
  'clientsecret',
  'jwt',
  'bearer',
  'credential',
  'credentials',
  // Plurals name the same secrets, often a whole list of them (narduk-libs#872).
  'tokens',
  'passwords',
  'passwds',
  'secrets',
  'apikeys',
  'jwts',
  'privatekeys',
  'clientsecrets',
])
/** Adjacent segments that together name a secret (`x-api-key` → api+key). */
const COMPOUND_SEGMENTS = new Set([
  'apikey',
  'accesskey',
  'privatekey',
  'clientsecret',
  'setcookie',
])
/** Keep infix on the punctuation-stripped key only for these compounds. */
const SENSITIVE_INFIX = ['apikey', 'accesskey', 'privatekey', 'authorization'] as const
/**
 * Suffix match on the punctuation-stripped key. Segment splitting cannot see a
 * boundary in an all-lowercase concatenation such as `refreshtoken` or
 * `dbpassword`, so without this the narrowing would stop redacting names the
 * suffix matcher already covered. `tokenizer` / `secretary` / `jwtid` do not
 * end in these words, and `tokencount` / `passwordless` are carved out below.
 * The plural forms catch `refreshtokens` / `dbpasswords` (narduk-libs#872).
 */
const SENSITIVE_SUFFIXES = [
  'token',
  'password',
  'secret',
  'tokens',
  'passwords',
  'secrets',
] as const
/**
 * Metric / method flags that contain `token`, `password`, or `auth` but are not secrets.
 * The `*tokens` entries are the LLM usage counts providers report; they are numbers,
 * and redacting them would blind AI cost and latency logs once plurals redact (#872).
 */
const SAFE_NORMALIZED_KEYS = new Set([
  'tokencount',
  'passwordless',
  'authmethod',
  'authbackend',
  'authprovider',
  'inputtokens',
  'outputtokens',
  'prompttokens',
  'completiontokens',
  'totaltokens',
  'maxtokens',
  'reasoningtokens',
  'cachereadinputtokens',
  'cachecreationinputtokens',
])

/** Mark a value as private. It is redacted before any sink sees the record. */
export function privateValue(value: unknown): object {
  return Object.freeze({ [PRIVATE]: true, value })
}

export function cleanText(value: string, length = MAX_STRING): string {
  // eslint-disable-next-line no-control-regex -- Strip terminal and line controls before every sink.
  return value.slice(0, length).replaceAll(/[\u0000-\u001f\u007f-\u009f]/g, ' ')
}

function normalizeKey(key: string): string {
  return key.toLowerCase().replaceAll(/[^a-z0-9]/g, '')
}

/** Split `authToken`, `access_token`, and `x-api-key` on the original key. */
function keySegments(key: string): string[] {
  const pieces = key.split(/[^A-Za-z0-9]+/u).filter(Boolean)
  const segments: string[] = []
  for (const piece of pieces) {
    const parts = piece.split(/(?<=[a-z0-9])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])/u)
    for (const part of parts) {
      if (part) segments.push(part.toLowerCase())
    }
  }
  return segments
}

export function isSensitiveKey(key: string, extra: readonly string[] = []): boolean {
  const normalized = normalizeKey(key)
  if (SENSITIVE.has(normalized) || extra.some((item) => normalizeKey(item) === normalized)) {
    return true
  }
  if (!normalized || SAFE_NORMALIZED_KEYS.has(normalized)) return false
  const segments = keySegments(key)
  if (segments.some((segment) => SENSITIVE_SEGMENTS.has(segment))) return true
  for (let index = 0; index < segments.length - 1; index++) {
    if (COMPOUND_SEGMENTS.has(`${segments[index]}${segments[index + 1]}`)) return true
  }
  if (segments.includes('auth')) return true
  if (SENSITIVE_SUFFIXES.some((part) => normalized.endsWith(part))) return true
  return SENSITIVE_INFIX.some((part) => normalized.includes(part))
}

/** Malformed URLs never fall back to the original query/credential-bearing string. */
export function sanitizeUrlForLog(value: string): string {
  if (!value) return '[empty]'
  try {
    const url =
      value.startsWith('/') && !value.startsWith('//')
        ? new URL(value, 'https://relative.invalid')
        : new URL(value)
    if (!['http:', 'https:'].includes(url.protocol)) return '[unsupported URL]'
    const path = cleanText(url.pathname, 512)
    return value.startsWith('/') ? path : `${url.protocol}//${url.host}${path}`.slice(0, 512)
  } catch {
    return '[invalid URL]'
  }
}

export interface SanitizerOptions {
  redact?: readonly string[]
  includeStack?: boolean
}

/** Reading data descriptors avoids executing application getters and toJSON hooks. */
function ownValues(value: object): Record<string, PropertyDescriptor> {
  try {
    return Object.getOwnPropertyDescriptors(value)
  } catch {
    return {}
  }
}

function errorField(value: object, key: string): unknown {
  // Error.name is usually inherited. Walk only data descriptors, never accessors.
  let current: object | null = value
  for (let i = 0; current && i < 4; i++) {
    try {
      const descriptor = Object.getOwnPropertyDescriptor(current, key)
      if (descriptor) {
        if ('value' in descriptor) return descriptor.value
        // V8 exposes lazily formatted native Error stacks as accessors. Do not invoke user getters.
        if (
          key === 'stack' &&
          descriptor.get &&
          Function.prototype.toString.call(descriptor.get).includes('[native code]')
        ) {
          return descriptor.get.call(value)
        }
        return undefined
      }
      current = Object.getPrototypeOf(current) as object | null
    } catch {
      return undefined
    }
  }
  return undefined
}

function describeError(
  value: unknown,
  options: SanitizerOptions,
  seen: WeakSet<object>,
  depth: number,
): LogError {
  if (typeof value !== 'object' || value === null) {
    return {
      name: 'Error',
      message: typeof value === 'string' ? cleanText(value) : 'Non-error thrown',
    }
  }
  if (seen.has(value) || depth >= MAX_DEPTH)
    return { name: 'Error', message: '[Circular or truncated cause]' }
  seen.add(value)
  const name = errorField(value, 'name')
  const message = errorField(value, 'message')
  const code = errorField(value, 'code')
  const stack = errorField(value, 'stack')
  const cause = errorField(value, 'cause')
  const result: LogError = {
    name: typeof name === 'string' ? cleanText(name, 128) : 'Error',
    message: typeof message === 'string' ? cleanText(message) : 'Non-error thrown',
  }
  if (typeof code === 'string' || typeof code === 'number')
    result.code = cleanText(String(code), 128)
  if (options.includeStack && typeof stack === 'string') result.stack = cleanText(stack, 4096)
  if (cause !== undefined) result.cause = describeError(cause, options, seen, depth + 1)
  return result
}

export function sanitizeErrorForLog(value: unknown, options: SanitizerOptions = {}): LogError {
  try {
    return describeError(value, options, new WeakSet(), 0)
  } catch {
    return { name: 'Error', message: '[Unserializable error]' }
  }
}

interface WalkState {
  seen: WeakSet<object>
  nodes: number
  options: SanitizerOptions
}

function walk(value: unknown, state: WalkState, depth: number, key = ''): JsonValue {
  if (isSensitiveKey(key, state.options.redact)) return REDACTED
  if (++state.nodes > MAX_NODES || depth > MAX_DEPTH) return '[Truncated]'
  if (value === null || typeof value === 'boolean') return value
  if (typeof value === 'number') return Number.isFinite(value) ? value : String(value)
  if (typeof value === 'bigint') return String(value)
  if (typeof value === 'string') {
    return /(?:url|uri)$/i.test(key) ? sanitizeUrlForLog(value) : cleanText(value)
  }
  if (typeof value !== 'object') return `[${typeof value}]`
  if (state.seen.has(value)) return '[Circular]'
  const descriptors = ownValues(value)
  if (Object.getOwnPropertyDescriptor(value, PRIVATE)?.value === true) return REDACTED
  if (value instanceof Error) return sanitizeErrorForLog(value, state.options)
  if (value instanceof Date) {
    try {
      return Date.prototype.toISOString.call(value)
    } catch {
      return '[Invalid Date]'
    }
  }
  state.seen.add(value)
  if (Array.isArray(value)) {
    const result: JsonValue[] = []
    const length = Math.min(value.length, MAX_FIELDS)
    for (let i = 0; i < length; i++) {
      const descriptor = descriptors[String(i)]
      result.push(
        descriptor && 'value' in descriptor
          ? walk(descriptor.value, state, depth + 1, key)
          : '[Accessor]',
      )
    }
    if (value.length > MAX_FIELDS) result.push('[Truncated]')
    return result
  }
  const result: SafeFields = {}
  for (const [field, descriptor] of Object.entries(descriptors).slice(0, MAX_FIELDS)) {
    if (!descriptor.enumerable || ['__proto__', 'prototype', 'constructor'].includes(field))
      continue
    const safeKey = cleanText(field, 128)
    result[safeKey] =
      'value' in descriptor ? walk(descriptor.value, state, depth + 1, field) : '[Accessor]'
  }
  return result
}

export function sanitizeFields(value: unknown, options: SanitizerOptions = {}): SafeFields {
  try {
    const result = walk(value, { seen: new WeakSet(), nodes: 0, options }, 0)
    return result && typeof result === 'object' && !Array.isArray(result)
      ? result
      : { value: result }
  } catch {
    return { serializationError: '[Unserializable data]' }
  }
}

export function recordBytes(record: Readonly<LogRecord>): number {
  return encoder.encode(JSON.stringify(record)).byteLength
}

export function boundRecord(record: LogRecord): LogRecord {
  if (recordBytes(record) <= MAX_RECORD_BYTES) return record
  const bounded: LogRecord = { ...record, data: { truncated: true } }
  if (record.error) bounded.error = { name: record.error.name, message: record.error.message }
  return bounded
}

export function freezeRecord(record: LogRecord): Readonly<LogRecord> {
  const freeze = (value: unknown): void => {
    if (!value || typeof value !== 'object' || Object.isFrozen(value)) return
    for (const child of Object.values(value)) freeze(child)
    Object.freeze(value)
  }
  freeze(record)
  return record
}

File: packages/modules/narduk-logging/go/sanitize.go
package narduklogging

import (
	"errors"
	"fmt"
	"math"
	"net/url"
	"reflect"
	"strconv"
	"strings"
	"time"
	"unicode"
	"unicode/utf8"
)

const (
	maxRecordBytes  = 16 * 1024
	maxString       = 2048
	maxFields       = 50
	maxDepth        = 6
	maxNodes        = 500
	redacted        = "[REDACTED]"
	timestampLayout = "2006-01-02T15:04:05.000Z"
	jsMaxSafe       = 9007199254740991
)

var sensitive = map[string]struct{}{
	"password": {}, "passwd": {}, "secret": {}, "token": {}, "apikey": {},
	"authorization": {}, "proxyauthorization": {}, "cookie": {}, "cookies": {},
	"setcookie": {}, "session": {}, "sessionid": {}, "privatekey": {},
	"clientsecret": {}, "body": {}, "requestbody": {}, "responsebody": {},
	"payload": {}, "payment": {}, "cardnumber": {}, "cvv": {}, "email": {},
	"phone": {}, "address": {}, "latitude": {}, "longitude": {}, "prompt": {},
	"completion": {},
}

// Exact segments after camelCase / snake_case / kebab-case split.
var sensitiveSegments = map[string]struct{}{
	"password": {}, "passwd": {}, "secret": {}, "token": {}, "apikey": {},
	"authorization": {}, "cookie": {}, "cookies": {}, "setcookie": {},
	"session": {}, "sessionid": {}, "privatekey": {}, "clientsecret": {},
	"jwt": {}, "bearer": {}, "credential": {}, "credentials": {},
}

// Adjacent segments that together name a secret (`x-api-key` → api+key).
var compoundSegments = map[string]struct{}{
	"apikey": {}, "accesskey": {}, "privatekey": {}, "clientsecret": {}, "setcookie": {},
}

// Keep infix on the punctuation-stripped key only for these compounds.
var sensitiveInfix = []string{"apikey", "accesskey", "privatekey", "authorization"}

// Suffix match on the punctuation-stripped key. Segment splitting cannot see a
// boundary in an all-lowercase concatenation such as `refreshtoken` or
// `dbpassword`, so without this the narrowing would stop redacting names the
// suffix matcher already covered. `tokenizer` / `secretary` / `jwtid` do not
// end in these words, and `tokencount` / `passwordless` are carved out below.
var sensitiveSuffixes = []string{"token", "password", "secret"}

// Metric / method flags that contain `token`, `password`, or `auth` but are not secrets.
var safeNormalizedKeys = map[string]struct{}{
	"tokencount": {}, "passwordless": {}, "authmethod": {}, "authbackend": {},
	"authprovider": {},
}

var reservedSkip = map[string]struct{}{
	"__proto__": {}, "prototype": {}, "constructor": {},
}

// Private marks a value as private. It is redacted before any sink sees the record.
type Private struct{ Value any }

type walkState struct {
	redact       []string
	includeStack bool
	seen         map[uintptr]struct{}
	nodes        int
}

// CleanText strips terminal and line controls and bounds the string.
func CleanText(value string, limit int) string {
	return cleanText(value, limit)
}

func cleanText(value string, limit int) string {
	if limit < 0 {
		limit = 0
	}
	if utf8.RuneCountInString(value) > limit {
		value = string([]rune(value)[:limit])
	}
	var b strings.Builder
	b.Grow(len(value))
	for _, r := range value {
		if r <= 0x1f || (r >= 0x7f && r <= 0x9f) {
			b.WriteByte(' ')
		} else {
			b.WriteRune(r)
		}
	}
	return b.String()
}

func normalizeKey(key string) string {
	var b strings.Builder
	b.Grow(len(key))
	for _, r := range strings.ToLower(key) {
		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') {
			b.WriteRune(r)
		}
	}
	return b.String()
}

func keySegments(key string) []string {
	var pieces []string
	var current strings.Builder
	flush := func() {
		if current.Len() > 0 {
			pieces = append(pieces, current.String())
			current.Reset()
		}
	}
	for _, r := range key {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			current.WriteRune(r)
		} else {
			flush()
		}
	}
	flush()
	var segments []string
	for _, piece := range pieces {
		segments = append(segments, splitCamel(piece)...)
	}
	return segments
}

func splitCamel(piece string) []string {
	runes := []rune(piece)
	if len(runes) == 0 {
		return nil
	}
	var parts []string
	start := 0
	for i := 1; i < len(runes); i++ {
		prev, cur := runes[i-1], runes[i]
		split := isASCIILowerOrDigit(prev) && isASCIIUpper(cur)
		if !split && isASCIIUpper(prev) && isASCIIUpper(cur) && i+1 < len(runes) && isASCIILower(runes[i+1]) {
			split = true
		}
		if split {
			parts = append(parts, strings.ToLower(string(runes[start:i])))
			start = i
		}
	}
	parts = append(parts, strings.ToLower(string(runes[start:])))
	return parts
}

func isASCIILowerOrDigit(r rune) bool {
	return (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9')
}

func isASCIIUpper(r rune) bool { return r >= 'A' && r <= 'Z' }

func isASCIILower(r rune) bool { return r >= 'a' && r <= 'z' }

// IsSensitiveKey reports whether a field name is treated as a secret.
func IsSensitiveKey(key string, extra []string) bool {
	return isSensitiveKey(key, extra)
}

func isSensitiveKey(key string, extra []string) bool {
	normalized := normalizeKey(key)
	if _, ok := sensitive[normalized]; ok {
		return true
	}
	for _, item := range extra {
		if normalizeKey(item) == normalized {
			return true
		}
	}
	if normalized == "" {
		return false
	}
	if _, ok := safeNormalizedKeys[normalized]; ok {
		return false
	}
	segments := keySegments(key)
	for _, segment := range segments {
		if _, ok := sensitiveSegments[segment]; ok {
			return true
		}
	}
	for i := 0; i < len(segments)-1; i++ {
		if _, ok := compoundSegments[segments[i]+segments[i+1]]; ok {
			return true
		}
	}
	for _, segment := range segments {
		if segment == "auth" {
			return true
		}
	}
	for _, suffix := range sensitiveSuffixes {
		if strings.HasSuffix(normalized, suffix) {
			return true
		}
	}
	for _, part := range sensitiveInfix {
		if strings.Contains(normalized, part) {
			return true
		}
	}
	return false
}

// SanitizeURL strips credentials, query, and fragment. Malformed values never
// fall back to the original string.
func SanitizeURL(value string) string {
	return sanitizeURL(value)
}

func sanitizeURL(value string) string {
	if value == "" {
		return "[empty]"
	}
	if strings.HasPrefix(value, "/") && !strings.HasPrefix(value, "//") {
		parsed, err := url.Parse(value)
		if err != nil {
			return "[invalid URL]"
		}
		path := parsed.Path
		if path == "" {
			path = "/"
		}
		return cleanText(path, 512)
	}
	parsed, err := url.Parse(value)
	if err != nil {
		return "[invalid URL]"
	}
	if parsed.Scheme == "" {
		return "[invalid URL]"
	}
	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return "[unsupported URL]"
	}
	host := strings.ToLower(parsed.Hostname())
	if host == "" {
		return "[invalid URL]"
	}
	if strings.Contains(host, ":") {
		host = "[" + host + "]"
	}
	port := parsed.Port()
	if port != "" && !((parsed.Scheme == "http" && port == "80") || (parsed.Scheme == "https" && port == "443")) {
		host += ":" + port
	}
	path := parsed.Path
	if path == "" {
		path = "/"
	}
	return cleanText(parsed.Scheme+"://"+host+path, 512)
}

// SanitizeFields applies the shared privacy contract used by the TypeScript,
// Python, and Swift adapters.
func SanitizeFields(value any, redact []string) map[string]any {
	return sanitizeFields(value, redact, false)
}

func sanitizeFields(value any, redact []string, includeStack bool) (result map[string]any) {
	defer func() {
		if recover() != nil {
			result = map[string]any{"serializationError": "[Unserializable data]"}
		}
	}()
	state := &walkState{
		redact:       redact,
		includeStack: includeStack,
		seen:         map[uintptr]struct{}{},
	}
	walked := walk(value, 0, "", state)
	if m, ok := walked.(map[string]any); ok {
		return m
	}
	return map[string]any{"value": walked}
}

func walk(value any, depth int, key string, state *walkState) any {
	if isSensitiveKey(key, state.redact) {
		return redacted
	}
	if _, ok := value.(Private); ok {
		return redacted
	}
	state.nodes++
	if depth > maxDepth || state.nodes > maxNodes {
		return "[Truncated]"
	}
	if value == nil {
		return nil
	}
	switch v := value.(type) {
	case bool:
		return v
	case string:
		lower := strings.ToLower(key)
		if strings.HasSuffix(lower, "url") || strings.HasSuffix(lower, "uri") {
			return sanitizeURL(v)
		}
		return cleanText(v, maxString)
	case error:
		return sanitizeError(v, state.includeStack, map[uintptr]struct{}{}, 0)
	case time.Time:
		if v.IsZero() {
			return "[Invalid Date]"
		}
		return v.UTC().Format(timestampLayout)
	case int:
		return jsSafeInt(int64(v))
	case int8:
		return jsSafeInt(int64(v))
	case int16:
		return jsSafeInt(int64(v))
	case int32:
		return jsSafeInt(int64(v))
	case int64:
		return jsSafeInt(v)
	case uint:
		return jsSafeUint(uint64(v))
	case uint8:
		return jsSafeUint(uint64(v))
	case uint16:
		return jsSafeUint(uint64(v))
	case uint32:
		return jsSafeUint(uint64(v))
	case uint64:
		return jsSafeUint(v)
	case float32:
		return jsSafeFloat(float64(v))
	case float64:
		return jsSafeFloat(v)
	case []any:
		return walkSlice(v, depth, key, state)
	case map[string]any:
		return walkMap(v, depth, state)
	default:
		return walkReflect(value, depth, key, state)
	}
}

func walkSlice(values []any, depth int, key string, state *walkState) any {
	if p, ok := ptrOf(values); ok {
		if _, seen := state.seen[p]; seen {
			return "[Circular]"
		}
		state.seen[p] = struct{}{}
	}
	limit := len(values)
	if limit > maxFields {
		limit = maxFields
	}
	out := make([]any, 0, limit)
	for i := 0; i < limit; i++ {
		out = append(out, walk(values[i], depth+1, key, state))
	}
	if len(values) > maxFields {
		out = append(out, "[Truncated]")
	}
	return out
}

func walkMap(values map[string]any, depth int, state *walkState) any {
	if p, ok := ptrOf(values); ok {
		if _, seen := state.seen[p]; seen {
			return "[Circular]"
		}
		state.seen[p] = struct{}{}
	}
	out := map[string]any{}
	for field, child := range values {
		if len(out) >= maxFields {
			break
		}
		if _, skip := reservedSkip[field]; skip {
			continue
		}
		out[cleanText(field, 128)] = walk(child, depth+1, field, state)
	}
	return out
}

func walkReflect(value any, depth int, key string, state *walkState) any {
	rv := reflect.ValueOf(value)
	for rv.Kind() == reflect.Interface || rv.Kind() == reflect.Pointer {
		if rv.IsNil() {
			return nil
		}
		rv = rv.Elem()
	}
	switch rv.Kind() {
	case reflect.Slice, reflect.Array:
		if p, ok := ptrOf(rv.Interface()); ok {
			if _, seen := state.seen[p]; seen {
				return "[Circular]"
			}
			state.seen[p] = struct{}{}
		}
		length := rv.Len()
		limit := length
		if limit > maxFields {
			limit = maxFields
		}
		out := make([]any, 0, limit)
		for i := 0; i < limit; i++ {
			out = append(out, walk(rv.Index(i).Interface(), depth+1, key, state))
		}
		if length > maxFields {
			out = append(out, "[Truncated]")
		}
		return out
	case reflect.Map:
		if rv.Type().Key().Kind() != reflect.String {
			return "[Unsupported object]"
		}
		if p, ok := ptrOf(rv.Interface()); ok {
			if _, seen := state.seen[p]; seen {
				return "[Circular]"
			}
			state.seen[p] = struct{}{}
		}
		out := map[string]any{}
		for _, mapKey := range rv.MapKeys() {
			if len(out) >= maxFields {
				break
			}
			field := mapKey.String()
			if _, skip := reservedSkip[field]; skip {
				continue
			}
			out[cleanText(field, 128)] = walk(rv.MapIndex(mapKey).Interface(), depth+1, field, state)
		}
		return out
	case reflect.Bool:
		return rv.Bool()
	case reflect.String:
		return walk(rv.String(), depth, key, state)
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		return jsSafeInt(rv.Int())
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return jsSafeUint(rv.Uint())
	case reflect.Float32, reflect.Float64:
		return jsSafeFloat(rv.Float())
	default:
		return "[Unsupported object]"
	}
}

func ptrOf(value any) (uintptr, bool) {
	rv := reflect.ValueOf(value)
	switch rv.Kind() {
	case reflect.Map, reflect.Slice, reflect.Pointer:
		if rv.IsNil() {
			return 0, false
		}
		return rv.Pointer(), true
	default:
		return 0, false
	}
}

func jsSafeInt(value int64) any {
	if value > jsMaxSafe || value < -jsMaxSafe {
		return strconv.FormatInt(value, 10)
	}
	return value
}

func jsSafeUint(value uint64) any {
	if value > jsMaxSafe {
		return strconv.FormatUint(value, 10)
	}
	return value
}

func jsSafeFloat(value float64) any {
	if math.IsNaN(value) {
		return "nan"
	}
	if math.IsInf(value, 1) {
		return "inf"
	}
	if math.IsInf(value, -1) {
		return "-inf"
	}
	return value
}

func sanitizeError(err error, includeStack bool, seen map[uintptr]struct{}, depth int) map[string]any {
	if err == nil {
		return map[string]any{"name": "Error", "message": "Non-error thrown"}
	}
	if depth >= maxDepth {
		return map[string]any{"name": "Error", "message": "[Circular or truncated cause]"}
	}
	if p, ok := ptrOf(err); ok {
		if _, found := seen[p]; found {
			return map[string]any{"name": "Error", "message": "[Circular or truncated cause]"}
		}
		seen[p] = struct{}{}
	}
	name := errorName(err)
	message := cleanText(err.Error(), maxString)
	result := map[string]any{"name": cleanText(name, 256), "message": message}
	if includeStack {
		result["stack"] = cleanText(fmt.Sprintf("%+v", err), 4096)
	}
	if cause := errors.Unwrap(err); cause != nil {
		result["cause"] = sanitizeError(cause, includeStack, seen, depth+1)
	}
	return result
}

func errorName(err error) string {
	t := reflect.TypeOf(err)
	if t == nil {
		return "Error"
	}
	if t.Kind() == reflect.Pointer {
		t = t.Elem()
	}
	if t.Name() == "" {
		return "Error"
	}
	return t.Name()
}

func errorFields(value any, includeStack bool, depth int) map[string]any {
	if depth >= maxDepth {
		return map[string]any{"name": "Error", "message": "[Circular or truncated cause]"}
	}
	if err, ok := value.(error); ok {
		return sanitizeError(err, includeStack, map[uintptr]struct{}{}, depth)
	}
	fields, ok := value.(map[string]any)
	if !ok {
		if s, ok := value.(string); ok {
			return map[string]any{"name": "Error", "message": cleanText(s, maxString)}
		}
		return map[string]any{"name": "Error", "message": "Non-error thrown"}
	}
	result := map[string]any{"name": "Error", "message": "Non-error thrown"}
	if s, ok := fields["name"].(string); ok {
		result["name"] = cleanText(s, 256)
	}
	if s, ok := fields["message"].(string); ok {
		result["message"] = cleanText(s, maxString)
	}
	switch code := fields["code"].(type) {
	case string:
		result["code"] = cleanText(code, 128)
	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, float32, float64:
		result["code"] = cleanText(fmt.Sprint(code), 128)
	}
	if includeStack {
		if s, ok := fields["stack"].(string); ok {
			result["stack"] = cleanText(s, 4096)
		}
	}
	if cause, ok := fields["cause"]; ok {
		result["cause"] = errorFields(cause, includeStack, depth+1)
	}
	return result
}
